### ISSUE TITLE
Question: What is `internalRuntype` for?

### ISSUE THREAD

**Issue by cakoose** [hoeck/simple-runtypes#32]
Just browsing through the code, I noticed the `internalRuntype` thing.  I haven't looked to deeply, but my initial concern was that the library might be using mechanisms that weren't available to normal users like me.  Is that the case?

(I'm looking at various TS validation libraries.  I initially just picked Zod but that was a mistake!  I'm optimistic about simple-runtypes because it's strict by default and allows you to transform+validate together.)

**Comment by hoeck** [comment]
There are indeed a few optimisation mechanisms that are (currently) not exposed to end users:

1. `isPure` hidden attribute (set with the `internalRuntype` method) - avoids expensive copying of the input for runtypes that do not modify the input
2. `fields` and `literal` hidden attributes - used by combinator runtypes such as `pick`, `intersection`, `union` and to build a lookup table for tagged unions
3. `failOrThrow` hidden parameter - used to decide whether a runtype call is toplevel (throw an error) or nested (return a `Fail` object), avoids using expensive try-catch blocks internally

If it turns out that any of those has a good use case being exposed, I'm happy to provide a public API for them. But right know it seemed they are better left private. And also less stuff for me to document :D.

Using the custom `runtype` allows you to use a plain function to implement your runtype in any way. I use that for instance to verify buffers in node or to parse weird date formats.

TLDR: yes, but there are good reasons & I'm not dogmatic about it

**Comment by cakoose** [comment]
Thanks for the explanations!  (Initially I was just looking through the code for examples of how to define my own types but ran into the fact that all the built-in types use `internalRuntype`.)

> `isPure` hidden attribute (set with the `internalRuntype` method) - avoids expensive copying of the input for runtypes that do not modify the input

Oh, interesting.  What's the policy around modifying input?  Some policies I have seen:
- parsers must make a copy of any mutable input -- obviously the safest option
- parsers take ownership of their inputs, which allows them to use the input objects to construct the output object -- helps perf for the common case of doing "JSON.parse" and then handing the result over to runtypes.

> `failOrThrow` hidden parameter - used to decide whether a runtype call is toplevel (throw an error) or nested (return a `Fail` object), avoids using expensive try-catch blocks internally

What are user-defined runtypes expected do, return a `Fail` object or throw?

Maybe I'm missing something, but what about just having all runtypes be defined to return a `Fail` object?  And then callers who want an exception can call a function that converts the `Fail` into a thrown exception?

**Comment by hoeck** [comment]
Regarding input, I deal with it as if it were immutable and I'm using copy-on-write to minimize allocations (so the second policy that you describe).

For example, the basic `string` runtype only does a simple `typeof value === 'string'` and will return the string object unmodified. The `string` runtype with the `trim` option will return a copy of the string as it needs to modify it.

Constructing a `record` from such a string will also cause the record-object to be copied as it needs to include the modified string. But only fields in the record that belong to an immutable runtype are copied, again saving allocations.

> What are user-defined runtypes expected do, return a Fail object or throw?

User defined runtypes are expected to return a `Fail` as the type signature (hopefully) [suggests](https://github.com/hoeck/simple-runtypes/blob/1f2a9afa6b19c503bdebdcf82375edd31491c343/src/custom.ts#L22).

BTW, If you have ideas on how to improve the documentation .. :grinning: I'm all ears.

> Maybe I'm missing something, but what about just having all runtypes be defined to return a Fail object? And then callers who want an exception can call a function that converts the Fail into a thrown exception?

Really good question. If you look at the version number, I'm at v7, so 7 major changes to the API :laughing:. At one point (I think v5) I was exactly using that model: no throws, just return a `T | Fail` from each runtype.

The consequence was that whenever I was using my runtypes, I would have to add an additional import. So before my code looked like this:

```import * as apiTypes from '../apiTypes`

**Comment by cakoose** [comment]
So would it be correct to say: an `isPure` runtype is one that, on success, just returns the input value.  It checks the input but does not transform it.  Is there anything else to it?

Re: returning `Fail` vs throwing.  I can definitely sympathize with going back and forth on an API :-)  However, in our codebase we mostly avoid exceptions and use a success/fail discriminated union everywhere, so that's what my personal preference is too.

Anyway, thank you so much for explaining everything.  This has been very helpful!

**Comment by hoeck** [comment]
> So would it be correct to say: an isPure runtype is one that, on success, just returns the input value. It checks the input but does not transform it. Is there anything else to it?

No, there is nothing more to it. It is a performance hack to avoid allocations. Container runtypes (record, array) use the isPure attribute of their element-runtypes to decide whether the container itself must be copied and rebuild too or can be returned as is.

> Re: returning Fail vs throwing. I can definitely sympathize with going back and forth on an API :-) However, in our codebase we mostly avoid exceptions and use a success/fail discriminated union everywhere, so that's what my personal preference is too.

Understood. I use either approach, depending on the situation.

For instance, at the backend at work I use exceptions almost exclusively, all caught and logged by a Koa middleware.
But the frontend should obviously never crash hard so thats why I included the `st.use(runtype, value)` function, checking for fails manually and just logging errors to the console.

> Thank you so much for explaining everything. This has been very helpful!
You're welcome!
